modules/ts/src/ts_tags.cpp

```diff
@@ -0,0 +1,471 @@
+            for (size_t i = 0; i < tags.size(); i++)
+            {
+                os << (i == 0 ? "'" : ", '") << tags[i] << "'";
+                os_direct << (i == 0 ? "" : ",") << tags[i];
+            }
+            std::string tags_str = os.str();
+            if (!tags.empty())
+                std::cout << "TEST: Run tests with tags: " << tags_str << std::endl;
+            ::testing::Test::RecordProperty("test_tags", os_direct.str());
+        }
+        {
+            const std::vector<std::string>& tags = getTestTagsSkipList();
+            std::ostringstream os, os_direct;
+            for (size_t i = 0; i < tags.size(); i++)
+            {
+                os << (i == 0 ? "'" : ", '") << tags[i] << "'";
+                os_direct << (i == 0 ? "" : ",") << tags[i];
+            }
+            std::string tags_str = os.str();
+            if (!tags.empty())
+                std::cout << "TEST: Skip tests with tags: " << tags_str << std::endl;
+            ::testing::Test::RecordProperty("test_tags_skip", os_direct.str());
+        }
+        {
+            const std::vector<std::string>& tags = getTestTagsForceList();
+            std::ostringstream os, os_direct;
+            for (size_t i = 0; i < tags.size(); i++)
+            {
+                os << (i == 0 ? "'" : ", '") << tags[i] << "'";
+                os_direct << (i == 0 ? "" : ",") << tags[i];
+            }
+            std::string tags_str = os.str();
+            if (!tags.empty())
+                std::cout << "TEST: Force tests with tags: " << tags_str << std::endl;
+            ::testing::Test::RecordProperty("test_tags_force", os_direct.str());
+        }
+    }
+
+    void OnTestStart(const ::testing::TestInfo& test_info) CV_OVERRIDE
+    {
+        currentDirectTestTags.clear();
+        currentImpliedTestTags.clear();
+
+        const char* value_param_ = test_info.value_param();
+        if (value_param_)
+        {
+            std::string value_param(value_param_);
+            if (value_param.find("CV_64F") != std::string::npos
+                || (value_param.find("64F") != std::string::npos
+                    && value_param.find(" 64F") != std::string::npos
+                    && value_param.find(",64F") != std::string::npos
+                    && value_param.find("(64F") != std::string::npos
+                )
+            )
+                applyTestTag_(CV_TEST_TAG_TYPE_64F);
+            if (value_param.find("1280x720") != std::string::npos)
+                applyTestTag_(CV_TEST_TAG_SIZE_HD);
+            if (value_param.find("1920x1080") != std::string::npos)
+                applyTestTag_(CV_TEST_TAG_SIZE_FULLHD);
+            if (value_param.find("3840x2160") != std::string::npos)
+                applyTestTag_(CV_TEST_TAG_SIZE_4K);
+        }
+    }
+
+    void OnTestEnd(const ::testing::TestInfo& /*test_info*/) CV_OVERRIDE
+    {
+        if (currentDirectTestTags.empty() && currentImpliedTestTags.empty())
```

Comment:
What is the difference between direct and implied tags?

Reply:
`currentDirectTestTags` contains list of passed tags via `applyTestTag()` calls.

`currentImpliedTestTags` contains list of **additional** implied tags. They are not printed on stdout, but they are used for filtering.

If `currentDirectTestTags` is empty then `currentImpliedTestTags` should be empty too.